dartagnan/src/main/java/com/dat3m/dartagnan/wmm/analysis/RelationAnalysis.java

```diff
@@ -1196,48 +1208,52 @@ public Delta visitInverse(Relation rel, Relation r1) {
         public Delta visitTransitiveClosure(Relation rel, Relation r1) {
             if (r1.equals(source)) {
                 Set<Tuple> maySet = new HashSet<>(may);
-                Knowledge k = knowledgeMap.get(rel);
-                Map<Event, List<Event>> mayMap = map(k.may);
-                for (Collection<Tuple> current = may; !current.isEmpty(); ) {
-                    update(mayMap, current);
-                    Collection<Tuple> next = new HashSet<>();
-                    for (Tuple tuple : current) {
-                        Event e1 = tuple.getFirst();
-                        for (Event e2 : mayMap.getOrDefault(tuple.getSecond(), List.of())) {
-                            Tuple t = new Tuple(e1, e2);
-                            if (!k.containsMay(t) && !maySet.contains(t) && !exec.areMutuallyExclusive(e1, e2)) {
-                                next.add(t);
-                            }
+                computeTransitiveClosure(maySet, map(knowledgeMap.get(rel).may), map(may), true);
+                if (enableMustSets) {
+                    Set<Tuple> mustSet = new HashSet<>(must);
+                    computeTransitiveClosure(mustSet, map(knowledgeMap.get(rel).must), map(must), false);
+                    return new Delta(maySet, mustSet);
+                }
+                return new Delta(maySet, EMPTY_SET);
+            }
+            return EMPTY;
+        }
+
+        private void computeTransitiveClosure(Set<Tuple> result, Map<Event, Set<Event>> outer, Map<Event, Set<Event>> inner, boolean isMay) {
+            Map<Event, Set<Event>> next;
+            inner.forEach((k, v) -> outer.computeIfAbsent(k, key -> new HashSet<>()).addAll(v));
+            for (Map<Event, Set<Event>> current = inner; !current.isEmpty(); current = next) {
+                next = new HashMap<>();
+                for (Map.Entry<Event, Set<Event>> entry : current.entrySet()) {
+                    Event e1 = entry.getKey();
+                    Set<Event> update = new HashSet<>();
+                    for (Event e2 : entry.getValue()) {
+                        if (isMay) {
+                            update.addAll(outer.getOrDefault(e2, Set.of()));
+                        } else {
+                            boolean implies = exec.isImplied(e1, e2);
+                            update.addAll(outer.getOrDefault(e2, Set.of()).stream()
+                                    .filter(e -> implies || exec.isImplied(e, e2))
+                                    .collect(toSet()));
```

Comment:
Not sure how much impact this has, but you could either use `.collect(toList())` here (cheaper than set construction, faster iteration) or directly do `outer.getOrDefault(...).stream().filter().foreach(update::add)` which would skip the construction of an intermediate collection altogether.
Since the must-case is usually not the bottleneck, I don't expect much gain so you are free to keep it as is if you want.

Reply:
Changed to a list, though don't see a difference on "my" benchmarks.